libs/community/langchain_community/vectorstores/duckdb.py

```diff
@@ -0,0 +1,192 @@
+                self._text_key: text,
+                self._vector_key: embedding,
+                "metadata": metadata,
+            }
+            docs.append(doc)
+        df = pd.DataFrame(docs)
+        print(df['embedding'])
+        # self._table.insert(df)
+        self._connection.sql(f"INSERT INTO {self._table_name} SELECT * FROM df")
+        return ids
+
+    def similarity_search(self, query: str, k: int = 4, **kwargs: Any) -> List[Document]:
+        """Performs a similarity search for a given query string.
+
+        Args:
+            query: The query string to search for.
+            k: The number of similar texts to return.
+
+        Returns:
+            A list of Documents most similar to the query.
+        """
+        embedding = self._embedding.embed_query(query)  # type: ignore
+        list_cosine_similarity = self.duckdb.FunctionExpression('list_cosine_similarity', self.duckdb.ColumnExpression(self._vector_key), self.duckdb.ConstantExpression(embedding))
+        docs = (self._table.
+            select(*[self.duckdb.StarExpression(),list_cosine_similarity.alias("similarity")]).
+            order("similarity desc").
+            limit(k).
+            select(self.duckdb.StarExpression(exclude=["similarity", self._vector_key])).
+            fetchdf()
+        )
+        return [
+            Document(
+                page_content=docs[self._text_key][idx],
+                metadata=json.loads(docs["metadata"][idx]) if docs["metadata"][idx] else {},
+            )
+            for idx in range(len(docs))
+        ]
+
+    @classmethod
+    def from_texts(
+        cls,
+        texts: List[str],
+        embedding: Embeddings,
+        metadatas: Optional[List[dict]] = None,
+        connection: Any = None,
+        vector_key: Optional[str] = "vector",
+        id_key: Optional[str] = "id",
+        text_key: Optional[str] = "text",
+        table_name: Optional[str] = "embeddings",
+        **kwargs: Any,
+    ) -> DuckDB:
+        """Creates an instance of DuckDB and populates it with texts and their embeddings.
+
+        Args:
+            texts: List of strings to add to the vector store.
+            embedding: The embedding function or model to use for generating embeddings.
+            metadatas: Optional list of metadata dictionaries associated with the texts.
+            connection: DuckDB connection. If not provided, a new connection will be created.
+            vector_key: The column name for storing vectors. Defaults to "vector".
+            id_key: The column name for storing unique identifiers. Defaults to "id".
+            text_key: The column name for storing text. Defaults to "text".
+            table_name: The name of the table to use for storing embeddings. Defaults to "embeddings".
+        
+        Returns:
+            An instance of DuckDB with the provided texts and their embeddings added.
+        """
+        # Create an instance of DuckDB
+        instance = DuckDB(
```

Comment:
Important: The default instance needs to be configured properly for security

we should via:  autoload_known_extensions, autoinstall_known_extensions, disabled_filesystems

![image](https://github.com/langchain-ai/langchain/assets/3205522/6a9cc448-9711-4de4-b35b-18248f7e8fdf)

Reply:
I ended up setting and recommending `enable_external_access` to `false`, which seemed to be the most restricted option. I didn't set it for an existing connection, as this might interfere with the behaviour of the external code. 

I also mentioned it in the `security note`. 

Thanks for this recommendation.